oi/type_graph/Types.h

```diff
@@ -361,8 +360,10 @@ class Array : public Type {
 
   DECLARE_ACCEPT
 
-  virtual std::string name() const override {
-    return "OIArray<" + elementType_.name() + ", " + std::to_string(len_) + ">";
+  virtual const std::string& name() const override {
+    name_ = std::string{"OIArray<"} + elementType_.name() + ", " +
+            std::to_string(len_) + ">";
+    return name_;
```

Comment:
The ownership model and this field being mutable seems really suspicious. In a single threaded context, where constness makes much less sense anyway, this returned value is valid until the next call to `name()` - this is probably fine. In a multithreaded context, where mutable still applies, it's guaranteed to be valid never.

It seems like the copies are a lot safer to me. I looked at doing something clever with atomics to make this more explicit, but we'd still have no control over the lifetime of the return value. An alternative would be a reference counted string in an atomic that is only changed when necessary, which would guarantee the lifetime of the output with minimal heap movement.

Even putting a lock on this name wouldn't help much because you'd need to hold a read lock until the caller has finished using the result. I think the invariants of using name need to be documented if not codified, so at least if/when it bites us in future it won't be a complete surprise.

Reply:
I'll admit this is kind of hacky, and this approach mainly came about through laziness...

Maybe a better solution would be to have a `setName()` function which can be called once by NameGen, after which `name()` can just return a reference to a `const std::string`